File: DSA/Manav-Rajput/Day51_question1.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int left = 1;
        int right = getMax(nums);
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (computeSum(nums, mid, threshold) > threshold) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        
        return left;
    }
    
    private int getMax(int[] nums) {
        int max = nums[0];
        for (int num : nums) {
            if (num > max) {
                max = num;
            }
        }
        return max;
    }
    
    private int computeSum(int[] nums, int divisor, int threshold) {
        int sum = 0;
        for (int num : nums) {
            sum += (num + divisor - 1) / divisor;
            if (sum > threshold) {
                return sum; // Early exit if sum exceeds threshold
            }
        }
        return sum;
    }
}

File: DSA/Arshdeep-Kaur/Day 51/Day51_Question1.cpp
class Solution {
public:
    int findAnswer(vector<int> &nums, int val) {
        int sum = 0;
        for(int i=0; i<nums.size(); i++) {
            sum += ceil((double)(nums[i]) / (double)(val));
        }
        return sum;
    }

    int smallestDivisor(vector<int>& nums, int threshold) {
        int low = 1, high = *max_element(nums.begin(), nums.end());

        while(low <= high) {
            int mid = (low + high) / 2;
            int ans = findAnswer(nums, mid);
            if (ans <= threshold) {
                high = mid - 1;
            }
            else {
                low = mid+1;
            }
        }

        return low;
    }
};
